ouster-ros/src/os_pcap_node.cpp

```diff
@@ -0,0 +1,349 @@
+// clang-format off
+#include "ouster_ros/os_ros.h"
+// clang-format on
+
+#include <string>
+#include <thread>
+#include <chrono>
+#include <iomanip>
+
+#include "ouster_sensor_msgs/msg/packet_msg.h"
+#include "ouster_ros/os_sensor_node_base.h"
+#include "ouster_ros/visibility_control.h"
+
+#include "thread_safe_ring_buffer.h"
+#include <ouster/os_pcap.h>
+
+namespace sensor = ouster::sensor;
+using ouster::sensor_utils::PcapReader;
+using namespace std::chrono;
+
+using ouster_sensor_msgs::msg::PacketMsg;
+
+namespace ouster_ros {
+
+class OusterPcap : public OusterSensorNodeBase {
+   public:
+    OUSTER_ROS_PUBLIC
+    explicit OusterPcap(const rclcpp::NodeOptions& options)
+        : OusterSensorNodeBase("os_pcap", options) {
+        declare_parameters();
+    }
+
+    LifecycleNodeInterface::CallbackReturn on_configure(
+        const rclcpp_lifecycle::State&) {
+        RCLCPP_INFO(get_logger(), "on_configure() is called.");
+
+        try {
+            auto meta_file = get_meta_file();
+            auto pcap_file = get_pcap_file();
+            create_metadata_publisher();
+            load_metadata_from_file(meta_file);
+            open_pcap(pcap_file);
+            publish_metadata();
+            create_get_metadata_service();
+            RCLCPP_INFO(get_logger(), "Running in replay mode");
+        } catch (const std::exception& ex) {
+            RCLCPP_ERROR_STREAM(
+                get_logger(),
+                "exception thrown while configuring the sensor, details: "
+                    << ex.what());
+            // TODO: return ERROR on fatal errors, FAILURE otherwise
+            return LifecycleNodeInterface::CallbackReturn::ERROR;
+        }
+
+        return LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+    LifecycleNodeInterface::CallbackReturn on_activate(
+        const rclcpp_lifecycle::State& state) {
+        RCLCPP_DEBUG(get_logger(), "on_activate() is called.");
+        LifecycleNode::on_activate(state);
+        create_publishers();
+        if (imu_packet_pub) imu_packet_pub->on_activate();
+        if (lidar_packet_pub) lidar_packet_pub->on_activate();
+        allocate_buffers();
+        start_packet_processing_threads();
+        start_packet_read_thread();
+        return LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+    LifecycleNodeInterface::CallbackReturn on_error(
+        const rclcpp_lifecycle::State&) {
+        RCLCPP_DEBUG(get_logger(), "on_error() is called.");
+        // Always return failure for now
+        return LifecycleNodeInterface::CallbackReturn::FAILURE;
+    }
+
+    LifecycleNodeInterface::CallbackReturn on_deactivate(
+        const rclcpp_lifecycle::State& state) {
+        RCLCPP_DEBUG(get_logger(), "on_deactivate() is called.");
+        LifecycleNode::on_deactivate(state);
+        return LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+    LifecycleNodeInterface::CallbackReturn on_cleanup(
```

Comment:
Why does a pcap node have lifecycle?

Reply:
Because it inherits from OusterSensorNodeBase which is a LifecycleNode. The sensor and pcap replay nodes are similar and can use some refactoring in the future for sure. But for now I would leave as is.